dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp

```diff
@@ -388,15 +430,176 @@ bool shouldCompactStableWithTooMuchDataOutOfSegmentRange(const DMContext & conte
     seg->setValidDataRatioChecked();
     return (valid_rows < total_rows * (1 - invalid_data_ratio_threshold)) || (valid_bytes < total_bytes * (1 - invalid_data_ratio_threshold));
 }
+
 } // namespace GC
 
+SegmentPtr DeltaMergeStore::gcTrySegmentMerge(const DMContextPtr & dm_context, const SegmentPtr & segment)
+{
+    auto segment_rows = segment->getEstimatedRows();
+    auto segment_bytes = segment->getEstimatedBytes();
+    if (segment_rows >= dm_context->small_segment_rows || segment_bytes >= dm_context->small_segment_bytes)
+    {
+        LOG_FMT_TRACE(
+            log,
+            "GC - Merge skipped because current segment is not small, segment={} table={}",
+            segment->simpleInfo(),
+            table_name);
+        return {};
+    }
+
+    auto segments_to_merge = getMergeableSegments(dm_context, segment);
+    if (segments_to_merge.size() < 2)
+    {
+        LOG_FMT_TRACE(
+            log,
+            "GC - Merge skipped because cannot find adjacent segments to merge, segment={} table={}",
+            segment->simpleInfo(),
+            table_name);
+        return {};
+    }
+
+    LOG_FMT_DEBUG(
+        log,
+        "GC - Trigger Merge, segment={} table={}",
+        segment->simpleInfo(),
+        table_name);
+    auto new_segment = segmentMerge(*dm_context, segments_to_merge, false);
+    if (new_segment)
+    {
+        checkSegmentUpdate(dm_context, segment, ThreadType::BG_GC);
+    }
+
+    return new_segment;
+}
+
+SegmentPtr DeltaMergeStore::gcTrySegmentMergeDelta(const DMContextPtr & dm_context, const SegmentPtr & segment, const SegmentPtr & prev_segment, const SegmentPtr & next_segment, DB::Timestamp gc_safe_point)
+{
+    SegmentSnapshotPtr segment_snap;
+    {
+        std::shared_lock lock(read_write_mutex); // TODO: Do we really need this lock?
```

Comment:
This lock is needed because if we don't hold a lock on the segment map, the segment may be abandoned when before snapshot. And because there is no snapshot on the segment, the data of the segment(including DTFile and ColumnFile) will be deleted. So you cannot access these data through the snapshot. Maybe you can add some comment about it.(
And you also need to add a check like this here: https://github.com/pingcap/tiflash/blob/f8ab1f179d3273b579520e528ceecdcb79b76d5f/dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp#L52